### ISSUE TITLE
Flaky tests on big-endian board

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#113]
The first of the two big-endian boards on the Google buildbot here seems have a lot of flaky tests. (it has been good for the last 5 builds) http://build.chromium.org/p/client.v8/builders/V8%20Mips%20-%20big%20endian%20-%20nosnap%20-%201

I had seen the same flaky test a couple times (diff day) so I put those tests on the skip list. Now it flakes on other tests occasionally. 

I am concerned that the board itself is flaky, as we had not seen random failures here in Santa Clara. 

Note that this board runs no-snapshot, which we do not normally do.

Here are two examples of flakiness from today.

```
*** Example 1: fails during bootstrap???? ***

Test: cctest/test-run-jsops/BinopGreaterThan (flaky in a repeated run)
Flags: --turbo-asm --turbo-filter=* --always-opt
Command: out/Release/cctest --random-seed=1581529182 --turbo-asm --turbo-filter=* --always-opt test-run-jsops/BinopGreaterThan --nohard-abort --nodead-code-elimination --nofold-constants --testing_serialization_file=out/.serdes/serdes_BinopGreaterThan__turbo_asm__turbo_filter=*__always_opt

Run #1
Exit code: -11
Result: CRASH
Expected outcomes: PASS

Stderr:
Exception thrown during bootstrapping
Extension or internal compilation error.

Run #2
Exit code: 0
Result: PASS
Expected outcomes: PASS


----------------------------------------------------------------------------------------------------
*** Example 2: random hang ?? ***

Test: cctest/test-debug/RecursiveBreakpointsGlobal (flaky in a repeated run)
Flags: --turbo-asm --turbo-filter=* --always-opt
Command: out/Release/cctest --random-seed=-644736348 --turbo-asm --turbo-filter=* --always-opt test-debug/RecursiveBreakpointsGlobal --nohard-abort --nodead-code-elimination --nofold-constants --testing_serialization_file=out/.serdes/serdes_RecursiveBreakpointsGlobal__turbo_asm__turbo_filter=*__always_opt

Run #1
Exit code: -15
Result: TIMEOUT
Expected outcomes: FLAKY, PASS

Stdout:
BARRIER

BARRIER

BARRIER


Run #2
Exit code: 0
Result: PASS
Expected outcomes: FLAKY, PASS

Stdout:
BARRIER

BARRIER

BARRIER

BARRIER
```

**Comment by paul99** [comment]
Per discussion in slack #er8-board channel, Ákos is testing a board with multiple runs today. @kwan0xfff will try to setup a board later this week here in Santa Clara and run the tests. According to @MIPSbkirby we may also get our er8 boards back online on the BB soon.

All of this is to help make a decision if we should replace that board at Google, or if there is some software reason for the flakes. Let's try to keep remaining discussion on this topic here in this issue.

**Comment by paul99** [comment]
I also noticed 'profviz' test flaked on board 2. I saw same test flake once before. I _think_ this is one of the known-flaky profiler tests. Does anyone else know more info?

**Comment by paul99** [comment]
Is there anything new on this? I believe we were going to run looping tests on these boards, both at HJ and here in SC. @palfia , @kwan0xfff any updates on that testing?

Today, we had several hard fails and some flaky fails on mjsunit/unicodelctest & mjsunit/unicodelctest-no-optimizations. It seems strangely consistent. 

This looks _very_ bad for us on the public buildbot.

**Comment by paul99** [comment]
Dusan, could you please investigate these failures? I think we really need to solve this, as it's been flaky on the google bots for way too long.

Rick has 4 boards running here in Santa Clara, big-endian, no-snapshot. Looking at the logfiles, there are **two** main errors, and these seem to be same as the random/flaky errors on Google's bots here: 

```
http://build.chromium.org/p/client.v8/builders/V8%20Mips%20-%20big%20endian%20-%20nosnap%20-%201
http://build.chromium.org/p/client.v8/builders/V8%20Mips%20-%20big%20endian%20-%20nosnap%20-%202
```

Many tests fail with: `TypeError: Cannot read property 'length' of undefined`

This includes the very common fails on `mjsunit/unicodelctest` and `mjsunit/unicodelctest-no-optimization`

The 2nd common failure is: `Exception thrown during bootstrapping`. Could be same fail as above happening during bootstrap, since we do not run with snapshots.

```
From run-1208-1343-multi-runs.log:

=== mjsunit/function-without-prototype ===
Exception thrown during bootstrapping
Extension or internal compilation error.
Command: /export/mipssw052/exptv8x4/v8/out/mips.release/d8 --test --random-seed=97979087 --turbo-asm --turbo-filter=* --always-opt --nohard-abort --nodead-code-elimination --nofold-constants /export/mipssw052/exptv8x4/v8/test/mjsunit/mjsunit.js /export/mipssw052/exptv8x4/v8/test/mjsunit/function-without-prototype.js
exit code: -11
--- CRASHED ---

Similar errors in 
run-1207-1758-multi-runs.log:  === cctest/test-lockers/LockerUnlocker ===,  === mjsunit/regress/regress-346343 ===
```

There are also many timeouts from cctest/test-mark-compact/MarkCompactCollector. I am less concerned on this, as 1) I have not seen this fail on Google's bots, and 2) they are currently skipping this on ARM, due to a bug (https://codereview.chromium.org/764783003)

Rick is copying his 4 logfiles to Dropbox/mips/er8-eb-logs (these will be updated hourly via cron-job so you can observe any new errors that occur.)

**Comment by paul99** [comment]
Dear github, please send notifications to @palfia and @kwan0xfff 
For all our team: we do have some discussions on slack #er8-board channel. Quick discussions fine there, but please make sure critical info gets into this issue.

**Comment by paul99** [comment]
I've been debugging this unicodelctest.js and learned some interesting things, and got one big surprise.

I'v been running the tests on a loaded er8 board (running test suite concurrently), as Ákos suggested, and both debug and release builds fail fairly quickly (always less than 40 iterations in my test).

This reduced test case https://v8m.slack.com/files/paul/F0360N27L/uctest.js shows this failure:

```
$ out/mips.release/d8 --test --random-seed=-1618658303 --turbo-asm --turbo-filter=* --always-opt --nohard-abort --nodead-code-elimination --nofold-constants test/mjsunit/mjsunit.js uctest.js
starting...
0
Yowsah: it is a LoadIC call that is throwing undefined exception!!
uctest.js:14: TypeError: Cannot read property 'length' of undefined
    reout = re.test(s);
               ^
TypeError: Cannot read property 'length' of undefined
    at RegExp.test (native)
    at uctest.js:14:16
```

The reduced case shows these important points:
- It is _not_ related to the crazy-long regular expression, a short one will also fail.
- It is _not_ related to the large sparse arrays. A simple regexp shows the error. The big arrays are commented out.

I patched the only 2 spots in v8 that throw a TypeError for read property:

```
$ git diff
diff --git a/src/ic/ic.cc b/src/ic/ic.cc
index cfc2a81..bec6ab7 100644
--- a/src/ic/ic.cc
+++ b/src/ic/ic.cc
@@ -717,6 +717,7 @@ MaybeHandle<Object> LoadIC::Load(Handle<Object> object, Handle<Name> name) {
   // If the object is undefined or null it's illegal to try to get any
   // of its properties; throw a TypeError in that case.
   if (object->IsUndefined() || object->IsNull()) {
+    PrintF("Yowsah: it is a LoadIC call that is throwing undefined exception!!\n");
     return TypeError("non_object_property_load", object, name);
   }

diff --git a/src/runtime/runtime-object.cc b/src/runtime/runtime-object.cc
index 617644c..a5ff6ef 100644
--- a/src/runtime/runtime-object.cc
+++ b/src/runtime/runtime-object.cc
@@ -69,6 +69,7 @@ MaybeHandle<Object> Runtime::GetObjectProperty(Isolate* isolate,
                                                Handle<Object> object,
                                                Handle<Object> key) {
   if (object->IsUndefined() || object->IsNull()) {
+    PrintF("Wow: it is a RUNTIME call that is throwing undefined exception!!");
     Handle<Object> args[2] = {key, object};
     THROW_NEW_ERROR(isolate, NewTypeError("non_object_property_load",
                                           HandleVector(args, 2)),
```

We see from that it is a bad LoadIC. I was going to debug that, and I ran into a **very weird occurrence** that I have _not_ explained yet.

Early on I suspected that the `String.fromCharCode(i);` was the error, possibly sometimes returning undefined object leading to the regexp fail. So I added the line: 

```
assertTrue(s.length >= 0);
```

to reference the length property. It was all good, I assumed that the failure was in the regexp native code.

But while making the reduced test case, I made it short, 20 lines long. And then the debug output is different. You likely notice that the `uctest.js` file has 4k lines of **commented out** arrays. That gives the error shown above.

If you deleted everything past line 22 in the file, which is **all comments**, then the failure looks like this:

```
$ out/mips.release/d8 --test --random-seed=-1618658303 --turbo-asm --turbo-filter=* --always-opt --nohard-abort --nodead-code-elimination --nofold-constants test/mjsunit/mjsunit.js uctest.js 
starting...
0
1
2
Yowsah: it is a LoadIC call that is throwing undefined exception!!
uctest.js:11: TypeError: Cannot read property 'length' of undefined
    assertTrue(s.length >= 0);
                        ^
TypeError: Cannot read property 'length' of undefined
    at RegExp.test (native)
    at uctest.js:14:16
```

So the first printing of the TypeError says it is the assert of `s.length` and the 2nd printing says it is the `RegExp.test`. Even the line numbers are different. **Which one do we believe?**

I think this is some error to do with parsing the file, and associating file position to errors, but I still do not know which is right.

I would probably pursue `String.fromCharCode(i);` first, since (1) it would be easier to debug, and (2) I guess I think the error position might be more correct in a shorter file.

Please update this issue or slack with your findings. Maybe we can have a brief skype chat tomorrow.

**Comment by dusmil** [comment]
Nice inputs, very strange, I do not have theory that can explain this behavior.
Additionally I checked gc flags variations ( --stress-compaction, --gc-global), and disabling allocation from generated code( --noinline_new) and they do not make any difference. Also the test fail in the same exact way if regexp interpreted is used, therefore bug is not in RegExp engine, as Paul already proved by reducing regexp string.

**Comment by dusmil** [comment]
It would be very convenient to put Abort() in LoadIC for undefined object so we can examine stack or do back-trace.  I have already done this, but so far I did not encounter single failure in gdb.

**Comment by paul99** [comment]
I did a bunch of testing last week, and have some more data on this difficult bug. This will be a terse summary. Ask questions, and I can add more data.

This does not happen on ia32 (in over 3000 iterations of complete test suite), nor on arm Panda board, about 30 iterations of full suite. Both arch's were very perfectly consistent in the small number of tests they did fail. Both tested with no-snap. Both were overloaded with more parallel jobs than real cores on the CPUs.

Testing on er8 board, little-endian, running this uct3.js test:

```
// Copyright 2012 the V8 project authors. All rights reserved.

// Flags: --allow-natives-syntax

print("starting...");

var re = /^([a-j])$/;

var k = 0;
for (var j = 0; j < 90; j++ ) {
  print(j + '  ------------------------------------------------');
  for (var i = 0; i < 0x10000; i++) {
    var s = String.fromCharCode(i);
    // %DebugPrint(s);
    // print('----> ' + i + ' ' + s.length);
    assertTrue(s.length >= 1);
    // assertTrue(!!re.test(s) == !!answer[i]);
    reout = re.test(s);
    // print(reout);
    // boolified = !!reout;
    // assertTrue(boolified == !!answer[i]);
  }
}
```

Note lots of commented out lines, not much code left. I removed even the commented out unicode regex's, just in case they had some weird effect. They don't.

With this, test must be run with _both_ `no-snapshot` build _and_ `--always-opt` flag to fail. If I omit either, the test will always pass all 90 iterations (each one includes 0x10000 char conversions).

This test will fail in same circumstances on _single-core_ Sigma 8646 (74K cpu, 1 core, no VPE's, Linux sees it as a single-cpu). So it is not specifically a multi-core problem, though it is very possibly related to multiple-threads and/or processes:

If I run the test from one terminal, and in a second terminal start a second `d8` shell, that is very likely to provoke a TypeError on the first terminal running `uct3.js`. Since this is a whole separate process, it seems weird. Also note that the running uct3.js code is doing iterations, and _not_ restarting. It has completed its bootstrap at start, and from then is just running JS code. Something about the 2nd `d8` shell starting up can trigger a failure.

However, it is _not required_ to start a 2nd process. On both er8 & 8646, the `uct3.js` program will often/usually fail before 90 iterations are completed, when nothing else runs on the board. It just fails faster if something is run.

Another oddity: this code does _not_ look FP-related to me, but it runs a _lot slower_ on the er8 board (1GHz clock) than on the 8646 (800 MHz). (19 seconds on 8646, 110 seconds on er8). The only thing I can think of is FPU emulation. It is on my list to monitor this via debug_fs, but I have not done so yet.

**Comment by paul99** [comment]
From above data, I think this is a problem compiling the startup code with `--always-opt`. It most repeatably fails in these 2 unicode tests, reduced here to `uct3.js` above. Ákos showed that if we revert https://codereview.chromium.org/736003002 the unicode fails do not happen. 

But we still sometimes see it fail to bootstrap at all, or fail in some other tests. These are much more rare.

My conjecture is that this CL adds a more frequent tickle of whatever is broken in the startup code by `--always-opt`, but itself is not causing the fail.

However, this thought explains nothing about why error frequency is increased by running another shell process.

**Comment by paul99** [comment]
Well, we might have gotten lucky on this bug. I noticed that the mips `eb` boards on the Google buildbot have not failed for a while. I started checking. My first guess was that Dusan's fix for sin-cos did it. But that was not the case.

My pseudo-bisect now shows that this was fixed with 701c6e747 (https://codereview.chromium.org/882683002) Balázs port of `MIPS: Use a trampoline stub to load the type feedback vector for CallICs.`

So far, I have run 36 iterations of my uct7.js script without failure, each of which does a loop of 180 iterations. Here is the code, it is a small variation of the one above:

```
// Copyright 2012 the V8 project authors. All rights reserved.
// Flags: --allow-natives-syntax

function explode_in_ruins() {
    print("starting...");
    var re = /^([abc])$/;

    var k = 0;
    for (var j = 0; j < 180; j++ ) {
        print(j + '  ------------------------------------------------');
        for (var i = 0; i < 0x10000; i++) {
            var s = String.fromCharCode(i);
//          %DebugPrint(s);
            assertTrue(s.length >= 1);
            reout = re.test(s);
        }
    }
}

explode_in_ruins()
```

I run it like this: `out/mipsel.release/d8 --allow-natives-syntax --test --random-seed=-1618658303 --always-opt --nohard-abort --nodead-code-elimination --nofold-constants test/mjsunit/mjsunit.js uct7.js`

As mentioned above the bug only shows when using `--always-opt` in a `no-snapshot` build.

I run this on a Sigma 8646, it seems to run fastest that way. I have no trouble rapidly showing the errors with commits before this revision. Here is my log of testing, based on this Buildbot log from Google's bot: http://build.chromium.org/p/client.v8/builders/V8%20Mips%20-%20big%20endian%20-%20nosnap%20-%201?numbuilds=200

```
a manual-ish bisect

Jan 12 07:24    4b6316c5ff5e88bd38ffed453c70052be175eecd    failure #393   FAIL   Pass: 0, fail: 10

Jan 17 01:03    154f263561fa5729a824597ed20cad0ac721a83a    failure #422    Pass: 1, fail: 9

Jan 22 10:45    6f9d60cc6f78467ebc44300213c07dc21832d10b    failure #453    Pass: 0, fail: 10

Jan 28 23:48    d0f3e5302c8d89fee4995b86a2b83c9ecf5d1e7a    success #492    Pass: 10, fail: 0

Jan 26 08:35    0ae9eb038fd47f94493141ebf6144363363fac6b    success #473    Pass: 8, fail: 2   hmmmm, low freq fail??
    - same rev 2nd time: Pass: 10, fail: 0

Jan 23 10:02    d1e45d9b00c5c3620f80a7f32d7edb2c65cb36f7    success #463    Pass: 1, fail: 9

Jan 26 06:59    8e838f705c32c5a40e4618d47f8feaa4ca41757a    success #472    Pass: 1, fail: 9

Jan 27 20:33    aa609b546d392288f1403a954ab62c32722f657b    success #482    Pass: 10, fail: 0

Jan 27 00:34    44f1b9d1483b61da1677dc86347fb3dc0e64c68b    success #477    Pass: 23, fail: 7


Back to:
Jan 28 23:48    d0f3e5302c8d89fee4995b86a2b83c9ecf5d1e7a    success #492    Pass: 12, fail: 0

482 good, 477 bad
Jan 27 03:45    7f9b2fa699ae48d1a766e87c14f3b89a2c0d3cd0    success #479    Pass: 0, fail: 12

Jan 27 08:01    701c6e747569926c13160d668035ac05d363e5dc    failure #480    Pass: 12, fail: 0
   -- MIPS use trampoline stub
   -- ran this again: Pass: 24, fail: 0
```

Another note, regarding the log from google. My conclusion is that with build #480 we are OK. That build shows as a failure, but it is a bogus (infrastructure) failure, did not even run the tests. The tests _before_ 480 that failed typically have unicodelc or unicodelc-no-optimize, which is the most common of the user mode failures for this test. The failures _after_ #480 are something else, I believe they are related to a commit that we needed to port for MIPS, or for some other transient error.

I would request some help from Balazs to possibly regress this further. I was planning to do a test based on Michael's original CL 2a567faa https://codereview.chromium.org/871063002, with just your port of it cherry-picked on top, to eliminate other intermediate CL's.

I think we need to understand this to the details also, and from my initial look at this CL, I do not see how it fits in our other symptoms.

(while typing this, another 26 iterations have completed, error-free)

**Comment by paul99** [comment]
If it is of help, I use a simple bash script to run the tests multiple iterations:

```
$ cat ~/bin/repeat-for.sh 
#!/bin/bash

LOOPS=$1
pass=0
fail=0

shift

echo "I will run: $*"
for((n=1; n<=$LOOPS; n++)) 
{
  echo  "Iteration $n"
  $*
  if [ $? -eq 0 ] ; then ((pass+=1))
  else ((fail+=1))
  fi
}

echo "Pass: $pass, fail: $fail"
```

invoked by this for 30 runs: `repeat-for.sh 30 out/mipsel.release/d8 --allow-natives-syntax --test --random-seed=-1618658303 --always-opt --nohard-abort --nodead-code-elimination --nofold-constants test/mjsunit/mjsunit.js uct7.js`

**Comment by paul99** [comment]
I ran 300 iterations with 0 fails at rev 701c6e7475. Today I ran another 300 iterations / 0 fails on today's top-of-tree commit: a0d0c433. (snapshot=off)

We are getting 3 boards set up to do looping tests of check test suite, run without snapshots. This will be 2 big-endian boards and 1 little-endian.

**Comment by paul99** [comment]
The CL that fixed this issue is  `e5a082f0 Reland "Only use FreeSpace objects in the free list"`
https://codereview.chromium.org/882633002 

Please take a look to see if you see any clues as to the details of how it fixes it. There are certainly changes to startup code, and serialization a bit. I do not see any connection to --always-opt yet.

As I suspected, the actual fixing CL is between the Michael Stanton's `2a567faa Use a trampoline stub to load the type feedback vector for CallICs.` and Balázs' port of it: `701c6e74 MIPS: Use a trampoline stub to load the type feedback vector for CallICs.`

I checked out at 2a567faa, cherry-picked 701c6e74, as my baseline, and then merged in remaining commits for testing.

22421bbe still had some errors. 

Merging e5a082f has now passed 160 iterations, 0 failures. I will start a longer test overnight.

**Comment by dusmil** [comment]
This is great news! I confirmed it. By reverting the patch I can reproduce the bug easily.
This is a GC bug after all. I am wondering how we did not see failures on ARM.

**Comment by paul99** [comment]
`e5a082f0 Reland "Only use FreeSpace objects in the free list"` (as described above): Pass: 600, fail: 0

**Comment by paul99** [comment]
OK, **the good news** is that there is no evidence of this recurring on ERpro-8 boards little or big endian. I just went thru logs of running full test suites in loops over multiple days built with no-snapshot, and there were none of these issues.

But, **the bad news** is that in mips64el builds running on hardware (which _do_ have snapshots enabled) we are getting unicodelctest and related failures: `TypeError: Cannot read property 'length' of undefined`

I am going to close this issue, since the error originally documented here  does seem to be gone. 

I have opened https://github.com/paul99/v8m-rb/issues/124 for this mips64 failure mode.
